runtime/bcverify/rtverify.c

```diff
@@ -1775,14 +1775,14 @@ verifyBytecodes (J9BytecodeVerificationData * verifyData)
 						goto _inconsistentStack2;
 					}
 				} else {
-					/* Need to ensure that there is at least an Object reference on the stack for the 
-					 * invokeinterface receiver.  If the top of stack is a base type or TOP, then 
-					 * throw a verify error.  The check for the receiver to be an interface occurs in
-					 * the invokeinterface bytecode.
-					 * Note: we need to check whether the Object reference on the stack is initialized
-					 * so as to stop an uninitialized object from being addressed here by invokeinterface.
+					/* Throw a verify error for any of the following invokeinterface scenarios:
+					 * 1. The top of the stack holds a base type or TOP
+					 * 2. The top of the stack holds an object array
+					 * 3. The Object reference on the stack is uninitialized
+					 * The check for the receiver to be an interface occurs in the bytecode interpreter.
 					 */
 					if ((BCV_TAG_BASE_TYPE_OR_TOP == (type & BCV_TAG_MASK))
+						|| (BCV_OBJECT_OR_ARRAY == (type & BCV_TAG_MASK) && (type & BCV_ARITY_MASK))
```

Comment:
This doesn't seem to be correct in that `BCV_OBJECT_OR_ARRAY` stands for both an object reference (which is allowed for `JBinvokeinterface` and an array (which is something to be captured there).  In such case, you might have to check `BCV_ARITY_FROM_TYPE(type)` (which is arity of the object reference) to see whether it is an array or not.
```
BCV_ARITY_FROM_TYPE(type) > 0
```

Reply:
Please see the output of RI:
```
Exception Details:
  Location:
    TestInterface.method1()V @7: invokeinterface
  Reason:
    Type '[LTestInterface$A;' (current frame, stack[0]) is not assignable to 'TestInterface$I' <-------
  Current Frame:
    bci: @7
    flags: { }
    locals: { '[LTestInterface$A;' }
    stack: { '[LTestInterface$A;' }
```
in which the array `[LTestInterface$A;` doesn't match the expected reference `TestInterface$I` on the top of the operand stack.

Reply:
My change does accomplish that but the way you've written it is much more clear, I will update this.

Reply:
> My change does accomplish that...

The problem with your code is that it also disallows the object reference (supported by `invokeinterface`) interface which breaks most of interface specific tests even though it captures the error with the object array.

Reply:
Isn't the type going to be an array if the arity is not 0?

Reply:
> Isn't the type going to be an array if the arity is not 0?

When arity is not zero, it is definitely an array. (what I meant is that your original changes incorrectly disallow the object reference. Support there is an object reference with 0 arity,  your changes will reject it as invalid (`type & BCV_ARITY_MASK` and `type & BCV_TAG_MASK` are 0) because `BCV_OBJECT_OR_ARRAY `(0) means both object and object array are essentially references in such case, in which case it  can't distinguish whether the type is an object or an object array)

So you don't have to check `(BCV_OBJECT_OR_ARRAY == (type & BCV_TAG_MASK)` given it is definitely a reference when it passes the 1st condition `BCV_TAG_BASE_TYPE_OR_TOP == (type & BCV_TAG_MASK)` (error will be triggered if it is a primitive type or placeholder (top)).

In addition,  it's better to use `BCV_ARITY_FROM_TYPE(type)` (the existing macro) to check the arity instead of directly check with `BCV_TAG_MASK` which leaves the arity bits on the upper end.

Reply:
Okay thanks I see what you mean.